src/core/fixkey_hash_table.h

```diff
@@ -41,18 +41,17 @@ hash value, or because we fake it and explicitly store the hash value.)
 Not all the optimisations described above are in place yet. Starting with
 "minimum viable product", with a design that should support adding them.
 
-Also starting out by using two memory blocks, so that ASAN and valgrind can
-spot (some) problems.
-
```

Comment:
I had assumed this when I wrote the comment when I started out, but in the process of getting to the single block I realised that it felt slightly more efficient to have the hash entries reversed in memory (so that the zero index entry is highest, and the zero index metadata is lowest). And this then got us to the proper sandwich, with the control structure in the middle, and **everything** in one block. (Which had not been my planned design when I started. It's better than the original plan,)

But the upshot of the reverse/forwards setup and (hopefully) careful use of *un*signed variables for indexes, is that all out of range indexes fall *out*side the end of the allocated single block - metadata falls off the top, entries fall off the bottom. So ASAN and valgrind can spot both. We don't have the situation that a "naughty" read of an entry that is "too high" is in the metadata memory, and hence ASAN/valgrind don't fault it. Hence this comment became irrelevant (and maybe we could have gone with a single block earlier).

Reply:
Very nice!